Data Explorer: Use concurrent.futures background job queue to return get_column_profiles results asynchronously

wesm opened the pull request:
As discussed in #4300, comm requests have to be served in the order they are received. Since the data explorer has some request types, like `get_column_profiles` that can grow very expensive for large datasets, this blocks "fast" requests like `get_data_values` from executing while these "slow" requests are pending.

This change adds a `return_column_profiles` frontend method that allows these requests to be fulfilled in the background, allowing fast requests to be served unimpeded. This resolves the immediate performance issue that the data explorer is facing, so this pattern can be refined over time as we have need to do other asynchronous / background request handling in the kernels.

Here's what the "blocked request" behavior looks like on the main branch:

https://github.com/user-attachments/assets/e614bb8b-c265-4daa-9e15-4cd93fd5e671

And here is with this change:

https://github.com/user-attachments/assets/dce1eb8b-07d4-470a-b12c-c1f16566b8f2

The main difference is that the grid values load immediately, and then the null percentages load later when the `return_column_profiles` event is fired.

This change can't be merged until the corresponding changes are implemented in Ark.

wesm committed:
Refactor to use concurrent.futures background job queue

dfalbel reviewed (dismissed):
LGTM!

seeM commented on extensions/positron-python/python_files/positron/positron_ipykernel/utils.py:
```diff
@@ -241,6 +243,48 @@ async def cancel_tasks(tasks: Set[asyncio.Task]) -> None:
+        with self.lock:
+            futures = list(self.pending_futures)
+
+        for future in futures:
```
There's also `concurrent.futures.wait` ([docs](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait)) if we need more control in future.

seeM commented on src/vs/workbench/services/languageRuntime/common/languageRuntimeDataExplorerClient.ts:
```diff
@@ -340,7 +356,30 @@ export class DataExplorerClientInstance extends Disposable {
+				const callbackId = generateUuid();
+				const promise = new DeferredPromise<Array<ColumnProfileResult>>();
+				this._asyncTasks.set(callbackId, promise);
+				await this._positronDataExplorerComm.getColumnProfiles(callbackId, profiles, this._profileFormatOptions);
```
Should we also catch errors here and reject the deferred promise immediately? It will timeout after 10 seconds in any case but could be helpful for the user to have immediate feedback.

dfalbel replied:
Tracked here: https://github.com/posit-dev/positron/issues/4559
Currently the expected behavior is to catch all error and return empty results instead. We will need additional UI to somehow display these errors.

seeM commented on src/vs/workbench/services/languageRuntime/common/languageRuntimeDataExplorerClient.ts:
```diff
@@ -100,6 +102,11 @@ export class DataExplorerClientInstance extends Disposable {
 	_profileFormatOptions: FormatOptions;
 
+	/**
+	 * Promises for asynchronous tasks requested of the backend.
```
```suggestion
	 * Promises for asynchronous tasks requested of the backend, keyed by callback ID.
```

seeM commented on src/vs/workbench/services/languageRuntime/common/languageRuntimeDataExplorerClient.ts:
```diff
@@ -340,7 +356,30 @@ export class DataExplorerClientInstance extends Disposable {
+				setTimeout(() => {
+					// If the promise has already been resolved, do nothing.
+					if (promise.isSettled) {
+						return;
```
Do we also need to clean up `_asyncTasks` here?

wesm replied:
In principle the key should have been deleted from `_asyncTasks` in the resolution of the promise

seeM reviewed (dismissed):
LGTM! Works as advertised.

Does Ark have this same issue? If not, I wonder if we should keep the old `get_column_profiles` method and make this PR's implementation a new method. That way we don't need to change anything in Ark, and when we support concurrency in the Python kernel we'll use the original method too.

wesm commented:
I’m not sure if this issue affects Ark or not. My understanding from what @lionel- told me is that they also want to serve requests in order in Ark also, so the asynchronous API here may be the preferred way. 

@seeM could you take charge of pushing this over the finish line (with whatever changes are necessary in Ark) since I’m OOO for the next couple weeks?

jmcphers commented on positron/comms/data_explorer-backend-openrpc.json:
```diff
@@ -265,9 +265,17 @@
+			"description": "Async request for a statistical summary or data profile for batch of columns",
 			"params": [
+				{
+					"name": "callback_id",
```
It makes me a little itchy to have to complicate the contract b/c of the limitations of the backend, but I don't see a better way than what you've done here.

dfalbel committed:
Improve comment

Co-authored-by: Wasim Lorgat <mwlorgat@gmail.com>
Signed-off-by: Daniel Falbel <dfalbel@gmail.com>

dfalbel committed:
Fix merge issue

dfalbel committed:
Bump ark to 0.1.131

dfalbel committed:
Do not allow sending events responses while handling RPC messages (#4617)

Fixes an issue while running viewing a small polars data frame. In this
case, the background tasks that return the null cournts would be
processed faster than the actual RPC message, leading to an issue where
the `send_event` is sent before `send_result`, but with the message id
from `send_result`, which cause positron to interpret the event and the
return value of the RPC message.

This PR makes sure `send_event` waits for for the RPC message to be
handled before sending the event results.

```
import polars as pl

from datetime import date
df = pl.DataFrame(
    {
        "foo": [1, 2, 3],
        "bar": [6.0, 7.0, 8.0],
        "ham": [date(2020, 1, 2), date(2021, 3, 4), date(2022, 5, 6)],
        "a": [None, 2, 3],
        "b": [0.5, None, 2.5],
        "c": [True, None, False],
    }
)


%view df
```

wesm commented on extensions/positron-python/python_files/positron/positron_ipykernel/tests/test_data_explorer.py:
```diff
@@ -295,13 +295,14 @@ def test_register_table_with_variable_path(de_service: DataExplorerService):
 
 
-def test_shutdown(de_service: DataExplorerService):
+def test_shutdown(de_service: DataExplorerService, event_loop):
```
This unused `event_loop` arg can be removed also

testlabauto committed:
retries for table reads

dfalbel committed:
Try to improve error message by isolating `tableData` variables

dfalbel committed:
Remove unused event loop arg

dfalbel committed:
Bump ark to v0.1.134

wesm commented:
Thanks for pushing this through!! This is a big UX improvement in the data explorer
